File: advent_of_code_2019/Day8.py
from collections import defaultdict

digits = [int(x) for x in open(
    '/Users/mahesh/Documents/Contests/advent_of_code_2019/input/day8').read().strip()]
rows = 6
cols = 25
layers = int(len(digits) / (rows * cols))
image = defaultdict(lambda: defaultdict(int))
for l in range(layers):
    for i in range(rows * cols):
        image[l][i] += digits[l * rows * cols + i]

fewest_zeros = 25 * 6 + 1
fewest_zeros_layer = 0
for l in range(layers):
    zeros = 0
    for i in range(len(image[l])):
        if(image[l][i] == 0):
            zeros += 1
    if zeros < fewest_zeros:
        fewest_zeros = zeros
        fewest_zeros_layer = l

ones = 0
twos = 0
for i in range(len(image[fewest_zeros_layer])):
    if image[fewest_zeros_layer][i] == 1:
        ones += 1
    elif image[fewest_zeros_layer][i] == 2:
        twos += 1

values = ()

print(ones * twos)

File: advent_of_code_2019/Day8.java
import Constants;

import java.util.*;
import java.io.*;

public class Day8 {
    public static void main(String[] args) throws IOException {
        Scanner in = new Scanner(new BufferedReader(new FileReader(Constants.FILE_PATH + "day8")));
        List<int[][]> image = new ArrayList<int[][]>();
        String input = in.nextLine();

        int rows = 6, cols = 25;
        int row = 0, col = 0, layer = 0;
        for (int i = 0; i < input.length(); i++) {
            if (image.size() <= layer)
                image.add(new int[rows][cols]);
            image.get(layer)[row][col] = Character.getNumericValue(input.charAt(i));
            col++;
            if (col == cols) {
                col = 0;
                row++;
            }
            if (row == rows) {
                row = 0;
                layer++;
            }
        }

        int fewestZeroDigits = Integer.MAX_VALUE;
        int fewestZeroDigitsLayer = -1;
        for (int i = 0; i < image.size(); i++) {
            int zeros = 0;
            for (int j = 0; j < image.get(i).length; j++)
                for (int k = 0; k < image.get(i)[j].length; k++)
                    if (image.get(i)[j][k] == 0)
                        zeros++;
            if (zeros < fewestZeroDigits) {
                fewestZeroDigits = zeros;
                fewestZeroDigitsLayer = i;
            }
        }

        int ones = 0, twos = 0;
        int[][] fewestLayer = image.get(fewestZeroDigitsLayer);
        for (int i = 0; i < fewestLayer.length; i++) {
            for (int j = 0; j < fewestLayer[i].length; j++) {
                if (fewestLayer[i][j] == 1)
                    ones++;
                else if (fewestLayer[i][j] == 2)
                    twos++;
            }
        }

        System.out.println(ones * twos);

        int[][] finalImage = new int[rows][cols];
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                int finalColor = -1;
                for (int[][] layerImage : image) {
                    if (layerImage[i][j] == 2)
                        continue;
                    finalColor = layerImage[i][j];
                    break;
                }
                finalImage[i][j] = finalColor;
            }
        }

        for (int i = 0; i < finalImage.length; i++) {
            for (int j = 0; j < finalImage[i].length; j++) {
                System.out.print(finalImage[i][j] == 1 ? "#" : " ");
            }
            System.out.println();
        }
    }
}